File: src/main/scala/akka/persistence/inmemory/util/UUIDs.java
package akka.persistence.inmemory.util;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.nio.charset.Charset;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

public class UUIDs {
    public static final long MIN_CLOCK_SEQ_AND_NODE = 0x8080808080808080L;
    public static final long CLOCK_SEQ_AND_NODE = makeClockSeqAndNode();
    public static final long START_EPOCH = makeEpoch();
    public static final AtomicLong lastTimestamp = new AtomicLong(0L);

    public static UUID startOf(long timestamp) {
        return new UUID(makeMSB(fromUnixTimestamp(timestamp)), MIN_CLOCK_SEQ_AND_NODE);
    }

    public static UUID timeBased() {
        return new UUID(makeMSB(getCurrentTimestamp()), CLOCK_SEQ_AND_NODE);
    }

    public static long makeClockSeqAndNode() {
        long clock = new Random(System.currentTimeMillis()).nextLong();
        long node = makeNode();

        long lsb = 0;
        lsb |= (clock & 0x0000000000003FFFL) << 48;
        lsb |= 0x8000000000000000L;
        lsb |= node;
        return lsb;
    }

    public static long unixTimestamp(UUID uuid) {
        if (uuid.version() != 1)
            throw new IllegalArgumentException(String.format("Can only retrieve the unix timestamp for version 1 uuid (provided version %d)", uuid.version()));

        long timestamp = uuid.timestamp();
        return (timestamp / 10000) + START_EPOCH;
    }

    public static long getCurrentTimestamp() {
        while (true) {
            long now = fromUnixTimestamp(System.currentTimeMillis());
            long last = lastTimestamp.get();
            if (now > last) {
                if (lastTimestamp.compareAndSet(last, now))
                    return now;
            } else {
                long lastMillis = millisOf(last);
                // If the clock went back in time, bail out
                if (millisOf(now) < millisOf(last))
                    return lastTimestamp.incrementAndGet();

                long candidate = last + 1;
                // If we've generated more than 10k uuid in that millisecond,
                // we restart the whole process until we get to the next millis.
                // Otherwise, we try use our candidate ... unless we've been
                // beaten by another thread in which case we try again.
                if (millisOf(candidate) == lastMillis && lastTimestamp.compareAndSet(last, candidate))
                    return candidate;
            }
        }
    }

    public static long makeEpoch() {
        // UUID v1 timestamp must be in 100-nanoseconds interval since 00:00:00.000 15 Oct 1582.
        Calendar c = Calendar.getInstance(TimeZone.getTimeZone("GMT-0"));
        c.set(Calendar.YEAR, 1582);
        c.set(Calendar.MONTH, Calendar.OCTOBER);
        c.set(Calendar.DAY_OF_MONTH, 15);
        c.set(Calendar.HOUR_OF_DAY, 0);
        c.set(Calendar.MINUTE, 0);
        c.set(Calendar.SECOND, 0);
        c.set(Calendar.MILLISECOND, 0);
        return c.getTimeInMillis();
    }

    public static long fromUnixTimestamp(long tstamp) {
        return (tstamp - START_EPOCH) * 10000;
    }

    public static long millisOf(long timestamp) {
        return timestamp / 10000;
    }

    public static long makeMSB(long timestamp) {
        long msb = 0L;
        msb |= (0x00000000ffffffffL & timestamp) << 32;
        msb |= (0x0000ffff00000000L & timestamp) >>> 16;
        msb |= (0x0fff000000000000L & timestamp) >>> 48;
        msb |= 0x0000000000001000L; // sets the version to 1.
        return msb;
    }

    public static Set<String> getAllLocalAddresses() {
        Set<String> allIps = new HashSet<String>();
        try {
            InetAddress localhost = InetAddress.getLocalHost();
            allIps.add(localhost.toString());
            // Also return the hostname if available, it won't hurt (this does a dns lookup, it's only done once at startup)
            allIps.add(localhost.getCanonicalHostName());
            InetAddress[] allMyIps = InetAddress.getAllByName(localhost.getCanonicalHostName());
            if (allMyIps != null) {
                for (int i = 0; i < allMyIps.length; i++)
                    allIps.add(allMyIps[i].toString());
            }
        } catch (UnknownHostException e) {
            // Ignore, we'll try the network interfaces anyway
        }

        try {
            Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces();
            if (en != null) {
                while (en.hasMoreElements()) {
                    Enumeration<InetAddress> enumIpAddr = en.nextElement().getInetAddresses();
                    while (enumIpAddr.hasMoreElements())
                        allIps.add(enumIpAddr.nextElement().toString());
                }
            }
        } catch (SocketException e) {
            // Ignore, if we've really got nothing so far, we'll throw an exception
        }

        return allIps;
    }

    private static long makeNode() {
        try {

            MessageDigest digest = MessageDigest.getInstance("MD5");
            for (String address : getAllLocalAddresses())
                update(digest, address);

            Properties props = System.getProperties();
            update(digest, props.getProperty("java.vendor"));
            update(digest, props.getProperty("java.vendor.url"));
            update(digest, props.getProperty("java.version"));
            update(digest, props.getProperty("os.arch"));
            update(digest, props.getProperty("os.name"));
            update(digest, props.getProperty("os.version"));

            byte[] hash = digest.digest();

            long node = 0;
            for (int i = 0; i < 6; i++)
                node |= (0x00000000000000ffL & (long) hash[i]) << (i * 8);
            // Since we don't use the mac address, the spec says that multicast
            // bit (least significant bit of the first byte of the node ID) must be 1.
            return node | 0x0000010000000000L;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public static void update(MessageDigest digest, String value) {
        if (value != null)
            digest.update(value.getBytes(Charset.forName("UTF-8")));
    }
}

File: src/main/scala/io/github/alstanchev/pekko/persistence/inmemory/util/UUIDs.scala
package io.github.alstanchev.pekko.persistence.inmemory.util

import java.util.UUID

import org.apache.pekko.persistence.query.TimeBasedUUID

import scala.compat.Platform
import scala.util.Random

object UUIDs {
  implicit val TimeBasedUUIDOrdering = new Ordering[TimeBasedUUID] {
    override def compare(x: TimeBasedUUID, y: TimeBasedUUID): Int = {
      val xuuid: UUID = x.value
      val yuuid: UUID = y.value
      val comparedByTimestamp: Int = xuuid.timestamp().compare(yuuid.timestamp())
      if (comparedByTimestamp == 0) {
        val comparedByClock: Int = xuuid.clockSequence().compare(yuuid.clockSequence())
        if (comparedByClock == 0) {
          xuuid.node().compare(yuuid.node())
        } else comparedByClock
      } else comparedByTimestamp
    }
  }

  final val MIN_CLOCK_SEQ_AND_NODE: Long = 0x8080808080808080L
  final val NODE: Long = 256475483242313L

  def startOf(timestamp: Long): UUID = {
    new UUID(makeMSB(UUIDUtil.fromUnixTimestamp(timestamp)), MIN_CLOCK_SEQ_AND_NODE)
  }

  def timeBased(): UUID = {
    new UUID(makeMSB(UUIDUtil.getCurrentTimestamp()), ClockSeqAndNode)
  }

  final val ClockSeqAndNode: Long = {
    val clock: Long = new Random(Platform.currentTime).nextLong()
    0L |
      (clock & 0x0000000000003FFFL) << 48 |
      0x8000000000000000L |
      NODE
  }

  def makeMSB(timestamp: Long): Long = {
    0L |
      (0x00000000ffffffffL & timestamp) << 32 |
      (0x0000ffff00000000L & timestamp) >>> 16 |
      (0x0fff000000000000L & timestamp) >>> 48 |
      0x0000000000001000L
  }

  def unixTimestamp(uuid: UUID): Long = {
    require(uuid.version() == 1, s"Can only retrieve the unix timestamp for version 1 uuid (provided version ${uuid.version})")
    val timestamp = uuid.timestamp()
    (timestamp / 10000) + UUIDUtil.START_EPOCH
  }
}
